pkg/controller/job/job_controller_test.go

```diff
@@ -1893,15 +1899,23 @@ func TestSyncPastDeadlineJobFinished(t *testing.T) {
 					break
 				}
 			}
-			manager.clock.Sleep(time.Second)
+			manager.clock.Sleep(2 * time.Second)
 			err = wait.Poll(200*time.Millisecond, 3*time.Second, func() (done bool, err error) {
+				_, err = manager.syncJob(ctx, jobKey)
+				if err != nil {
+					return false, err
+				}
 				j, err = clientset.BatchV1().Jobs(metav1.NamespaceDefault).Get(ctx, job.GetName(), metav1.GetOptions{})
 				if err != nil {
-					return false, nil
+					return false, err
 				}
 				if len(j.Status.Conditions) == 1 && j.Status.Conditions[0].Reason == "DeadlineExceeded" {
 					return true, nil
 				}
+				// Advance the clock again. The fact that in the first pass the flow got up to here but the
+				// condition was not set means that you have hit an edge condition. So advance the clock
+				// here before the controller.syncJob is called again in the next polling interval.
+				manager.clock.Sleep(200 * time.Millisecond)
```

Comment:
Is there any case this is needed? I'm wondering that since you induce `syncJob` the periodic polling may not be needed at all - the first poll should always detect the new condition. If this is the case I would suggest removing the periodic poll entirely.

Reply:
I have seen this happen a few times while testing things locally. Even after setting the active deadline seconds to 2 seconds and sleep to 2 second to advance the clock. 

```bash
 690   │ === RUN   TestSyncPastDeadlineJobFinished
 691   │ === RUN   TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set
 692   │ I1228 23:01:11.596824   28776 job_controller.go:199] Starting job controller
 693   │ I1228 23:01:11.596842   28776 shared_informer.go:314] Waiting for caches to sync for job
 694   │ I1228 23:01:11.596854   28776 shared_informer.go:321] Caches are synced for job
 695   │ I1228 23:01:11.597018   28776 job_controller.go:516] enqueueing job default/job1
 696   │ I1228 23:01:11.597305   28776 job_controller.go:516] enqueueing job default/job1
 697   │ I1228 23:01:11.597547   28776 job_controller.go:516] enqueueing job default/job1
 698   │     job_controller_test.go:1908: Job failed to enforce activeDeadlineSeconds configuration. Expected condition with Reason 'DeadlineExceeded' was not found in {[] 2022-12-28 23:01:13 +0530 IST <nil> 0 0 0  nil 0xc0003d818c}
 699   │ === RUN   TestSyncPastDeadlineJobFinished/New_job_created_with_start_time_being_set
 700   │ I1228 23:01:14.597776   28776 job_controller.go:516] enqueueing job default/job2
 701   │ I1228 23:01:14.597824   28776 job_controller.go:516] enqueueing job default/job1
 702   │ I1228 23:01:14.597854   28776 job_controller.go:516] enqueueing job default/job2
 703   │ I1228 23:01:14.597868   28776 event.go:294] "Event occurred" object="default/job1" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
 704   │ I1228 23:01:14.597878   28776 event.go:294] "Event occurred" object="default/job2" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
 705   │ --- FAIL: TestSyncPastDeadlineJobFinished (3.30s)
 706   │     --- FAIL: TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set (3.00s)
 707   │     --- PASS: TestSyncPastDeadlineJobFinished/New_job_created_with_start_time_being_set (0.20s)
 708   │ I1228 23:01:14.797990   28776 job_controller.go:213] Shutting down job controller
```

Reply:
Other option is to just advance the clock by `3s` and set the `ActiveDeadlineSeconds` to be `2s`. But that is not the right way to test I think

Reply:
```bash
=== RUN   TestSyncPastDeadlineJobFinished
=== RUN   TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set
I1228 23:17:17.430555   33419 job_controller.go:199] Starting job controller
I1228 23:17:17.430569   33419 shared_informer.go:314] Waiting for caches to sync for job
I1228 23:17:17.430580   33419 shared_informer.go:321] Caches are synced for job
I1228 23:17:17.430660   33419 job_controller.go:516] enqueueing job default/job1
I1228 23:17:17.430748   33419 job_controller.go:1344] "Timing Information" job="job1" duration="0s" allowed="2s" start="2022-12-28 23:17:17 +0530 IST" current="2022-12-28 23:17:17 +0530 IST"
I1228 23:17:17.431045   33419 job_controller.go:1344] "Timing Information" job="job1" duration="0s" allowed="2s" start="2022-12-28 23:17:20 +0530 IST" current="2022-12-28 23:17:20 +0530 IST"
I1228 23:17:17.431056   33419 job_controller.go:516] enqueueing job default/job1
I1228 23:17:17.431100   33419 job_controller.go:516] enqueueing job default/job1
I1228 23:17:17.431131   33419 job_controller.go:1344] "Timing Information" job="job1" duration="0s" allowed="2s" start="2022-12-28 23:17:20 +0530 IST" current="2022-12-28 23:17:20 +0530 IST"
I1228 23:17:17.431169   33419 job_controller.go:1344] "Timing Information" job="job1" duration="0s" allowed="2s" start="2022-12-28 23:17:20 +0530 IST" current="2022-12-28 23:17:20 +0530 IST"
    job_controller_test.go:1908: Job failed to enforce activeDeadlineSeconds configuration. Expected condition with Reason 'DeadlineExceeded' was not found in {[] 2022-12-28 23:17:20 +0530 IST <nil> 0 0 0  nil 0xc00078812c}
=== RUN   TestSyncPastDeadlineJobFinished/New_job_created_with_start_time_being_set
I1228 23:17:20.431293   33419 job_controller.go:516] enqueueing job default/job2
I1228 23:17:20.431313   33419 job_controller.go:1344] "Timing Information" job="job1" duration="3s" allowed="2s" start="2022-12-28 23:17:20 +0530 IST" current="2022-12-28 23:17:23 +0530 IST"
I1228 23:17:20.431372   33419 job_controller.go:1344] "Timing Information" job="job2" duration="4s" allowed="2s" start="2022-12-28 23:17:19 +0530 IST" current="2022-12-28 23:17:23 +0530 IST"
I1228 23:17:20.431379   33419 job_controller.go:516] enqueueing job default/job1
I1228 23:17:20.431419   33419 job_controller.go:516] enqueueing job default/job2
I1228 23:17:20.431417   33419 event.go:294] "Event occurred" object="default/job1" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
I1228 23:17:20.431432   33419 event.go:294] "Event occurred" object="default/job2" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
--- FAIL: TestSyncPastDeadlineJobFinished (3.30s)
    --- FAIL: TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set (3.00s)
    --- PASS: TestSyncPastDeadlineJobFinished/New_job_created_with_start_time_being_set (0.20s)
```

Reply:
I think the overall interval also needs to be much higher for polling. 3 second max poll is not good enough for some cases either it seems like. So I should bump that up to 5s may be ? @mimowo

Reply:
```bash
=== RUN   TestSyncPastDeadlineJobFinished
=== RUN   TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set
I1228 23:23:03.546069   35573 job_controller.go:199] Starting job controller
I1228 23:23:03.546095   35573 shared_informer.go:314] Waiting for caches to sync for job
I1228 23:23:03.546113   35573 shared_informer.go:321] Caches are synced for job
I1228 23:23:03.546228   35573 job_controller.go:516] enqueueing job default/job1
I1228 23:23:03.546568   35573 job_controller.go:516] enqueueing job default/job1
I1228 23:23:03.546705   35573 job_controller.go:516] enqueueing job default/job1
    job_controller_test.go:1908: Job failed to enforce activeDeadlineSeconds configuration. Expected condition with Reason 'DeadlineExceeded' was not found in {[] 2022-12-28 23:23:05 +0530 IST <nil> 0 0 0  nil 0xc000a144dc}
=== RUN   TestSyncPastDeadlineJobFinished/New_job_created_with_start_time_being_set
I1228 23:23:08.548244   35573 job_controller.go:516] enqueueing job default/job2
I1228 23:23:08.548359   35573 job_controller.go:516] enqueueing job default/job1
I1228 23:23:08.548421   35573 event.go:294] "Event occurred" object="default/job1" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
I1228 23:23:08.548432   35573 job_controller.go:516] enqueueing job default/job2
I1228 23:23:08.548453   35573 event.go:294] "Event occurred" object="default/job2" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
--- FAIL: TestSyncPastDeadlineJobFinished (5.30s)
    --- FAIL: TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set (5.00s)
    --- PASS: TestSyncPastDeadlineJobFinished/New_job_created_with_start_time_being_set (0.20s)
```

Even that is giving false values on loop test. The safest is to switch to using syncJob call in this PR itself. Otherwise, I don't think we will get rid of the flakes fully.

Reply:
It should be ok to bump the time to 5s.

Reply:
I see that we still need the periodic poll due to cache, but do we still need `manager.clock.Sleep(200 * time.Millisecond)` and if so - can you explain why.